### PR TITLE
Information

### PR DESCRIPTION
**Pull request by mmatera**
Finally, I could add the prefix operators ? and ?? with its standard mea ning in mma: ? F is equivalent to Definition[F] and ??F calls Information[F].
I also added the command Information, which for built in symbols, shows the **doc** string of the correspondent class, and for user symbols, looks to the message ::usage.

Something that is missing is to show inside Information the output of Definition, but for some reason it didn't work. Anyway, I think the basic functionality is implemented.

### CODE REVIEW

File: mathics/builtin/assignment.py [hunk 1]
```diff
@@ -792,10 +800,130 @@ def rhs(expr):
...
+def information_interpret_doc_string(ds):
+    # TODO: work a little bit the interpretation
+    ds = ds.split("\n")
+    out = ""
+    for s in ds:
+        if s[0:6]=="    >>":
+            # If we reach the test section, stop the loop.
+            break
+        out = out + "\n" + s
+    return out
```
Comment (sn6uv) [inline mathics/builtin/assignment.py]: This is not really a robust way to parse the docstrings. I think it's better to reuse the existing stuff in `mathics/doc`.

File: mathics/builtin/assignment.py [hunk 2]
```diff
@@ -792,10 +800,130 @@ def rhs(expr):
...
+   
+    >> F[x_] := x^2
+    >> F::usage = "F[x] evaluates the square of x."
+    >> H[F[y_,u_]]^:=F[H[y],u]
+    >> Information[F]
+     |F[x] evaluates the square of x
+     = Null
+    """
+    operator="??"
+    precedence=5001
```
Comment (sn6uv) [inline mathics/builtin/assignment.py]: Is this correct?

File: mathics/core/parser/operators.py [hunk 3]
```diff
@@ -17,6 +17,8 @@
     'Exists': 240,
     'NotExists': 240,
     'Not': 230,
+    'Information': 5001,
+    'Definition' : 5000,
```
Comment (sn6uv) [inline mathics/core/parser/operators.py]: What's the correct precedence of these operators?

File: mathics/core/parser/tokeniser.py [hunk 4]
```diff
@@ -206,6 +206,9 @@
     ('Therefore', r' \u2234 '),
     ('Because', r' \u2235 '),
     ('Backslash', r' \u2216 '),
+
+    ('Information', r'\?\?'),
+    ('Definition', r'\?'),
```
Comment (sn6uv) [inline mathics/core/parser/tokeniser.py]: conflicts with `PatternTest` token.

File: mathics/core/parser/tokeniser.py [hunk 5]
```diff
@@ -230,7 +233,7 @@
     '<': ['Get', 'StringJoin', 'LessEqual', 'Less'],
     '=': ['SameQ', 'UnsameQ', 'Equal', 'Unset', 'Set'],
     '>': ['PutAppend', 'Put', 'GreaterEqual', 'Greater'],
-    '?': ['PatternTest'],
+    '?': ['Information', 'Definition','PatternTest'],
```
Comment (sn6uv) [inline mathics/core/parser/tokeniser.py]: Rather than adding an additional token we should reuse the `PatternTest` token for `Information`.

File: mathics/main.py [hunk 6]
```diff
@@ -246,6 +249,22 @@ def main():
...
+                query = evaluation.parse_feeder(feeder)
+                if query is None:
+                    continue
+                evaluation.evaluate(query, timeout=settings.TIMEOUT)
+        except (KeyboardInterrupt):
+            print('\nKeyboardInterrupt')
+        
+        definitions.set_line_no(0)
+
+
```
Comment (sn6uv) [inline mathics/main.py]: Could you split this `initfile` stuff out into a separate PR please.

File: mathics/builtin/assignment.py [hunk 7]
```diff
@@ -9,8 +9,10 @@
 from mathics.builtin.base import (
     Builtin, BinaryOperator, PostfixOperator, PrefixOperator)
 from mathics.core.expression import (Expression, Symbol, valid_context_name,
-                                     system_symbols)
-from mathics.core.rules import Rule
+                                     system_symbols,String)
+
```
Comment (sn6uv) [inline mathics/builtin/assignment.py]: Can you fix the formatting here. Should be: `system_symbols, String`

File: mathics/builtin/assignment.py [hunk 8]
```diff
@@ -625,6 +627,11 @@ class Definition(Builtin):
     >> Definition[a]
      = a = 2
 
+    >> a = 2;
+    >> ? a
+     = a = 2
+
+
```
Comment (sn6uv) [inline mathics/builtin/assignment.py]: `Definition` is not related to `Information`:

- `?a` is `Information["a", Rule[LongForm, False]]`
- `??a` is `Information["a", Rule[LongForm, True]]`

File: mathics/builtin/inout.py [hunk 9]
```diff
@@ -1820,6 +1820,7 @@ def apply_tex(self, expr, evaluation):
             # Replace multiple newlines by a single one e.g. between asy-blocks
             tex = MULTI_NEWLINE_RE.sub('\n', tex)
 
+
```
Comment (sn6uv) [inline mathics/builtin/inout.py]: there are a few whitespace changes that don't need to be included.

**Review by sn6uv — DISMISSED** [review]

File: mathics/builtin/assignment.py (hunk 2 above)
Comment (mmatera) [inline mathics/builtin/assignment.py]: I do not know if there is an standard value for the precedence of these operators. I just choose a very large one, because I supposed that in the context where these operators make sense, they should be evaluated with very high priority.

File: mathics/builtin/inout.py (hunk 9 above)
Comment (mmatera) [inline mathics/builtin/inout.py]: I notice that, but I do not know how to ask git to revert just a file, or ignore this kind of changes.

File: mathics/builtin/assignment.py (hunk 7 above)
Comment (mmatera) [inline mathics/builtin/assignment.py]: OK

File: mathics/core/parser/tokeniser.py (hunk 4 above)
Comment (sn6uv) [inline mathics/core/parser/tokeniser.py]: My point is that `PatternTest` and `Information` need to share the ` \? ` token.

File: mathics/core/parser/tokeniser.py (hunk 5 above)
Comment (sn6uv) [inline mathics/core/parser/tokeniser.py]: So there are two tokens `('SingleQM',  ' \? ')` and `('DoubleQM', '  \?\? ')`. This part of the tokeniser looks at a single character and tries to determine what token will match.

Let's say the tokeniser sees `? next. The problem is we don't know which token to emit without looking at the next character. The next character is either `?` or it's something else. The solution is to first try matching `DoubleQM` and then try `SingleQM`.

Does that make sense?

File: mathics/builtin/assignment.py (hunk 2 above)
Comment (sn6uv) [inline mathics/builtin/assignment.py]: True. It looks like they don't have a builtin precedence. `Precedence[Information]` gives the default of 670. You can experimenting to compare with other operators. You can wrap things in `Hold` to investigate.

For example:
```
In[2]:= Hold[a ? b ; c] // FullForm
Out[2]//FullForm= Hold[CompoundExpression[PatternTest[a, b], c]]
```
so the precedence is higher than `CompoundExpression`.

File: mathics/builtin/assignment.py (hunk 8 above)
Comment (sn6uv) [inline mathics/builtin/assignment.py]: Which version of MMA are you using?
```
In[1]:= $Version
Out[1]= 11.0.1 for Linux x86 (64-bit) (September 21, 2016)

In[2]:= Hold[??a]
Out[2]= Hold[Information[a, LongForm -> True]]

In[3]:= Hold[?a]
Out[3]= Hold[Information[a, LongForm -> False]]
```

Comment (mmatera) [inline mathics/builtin/assignment.py]: I use the version 9, and Hold[?a] also returns Hold[Information[a, LongForm -> False]]
,  but the help says something different: https://reference.wolfram.com/language/ref/Definition.html

File: mathics/builtin/assignment.py (hunk 2 above)
Comment (mmatera) [inline mathics/builtin/assignment.py]: Indeed, it seems to have smaller priority than Times, but larger than +...

In[1]:= Hold[??a + b]//FullForm
Out[1]//FullForm= Hold[Times[Information["a", Rule[LongForm, True]], Plus[b]]]

In[2]:= Hold[??a *b]//FullForm
Out[2]//FullForm= Hold[Information["a * b", Rule[LongForm, True]]]

File: mathics/core/parser/tokeniser.py (hunk 5 above)
Comment (mmatera) [inline mathics/core/parser/tokeniser.py]: OK, but then how we should implement it? I tried to imitate the behavior of @@ and >>, in a way that the tokeniser first look at the double operator, and afterward to the simple one.  This is why the change at line 236. What I do not realise is how and when the tokeniser distinguishes a prefix operator from an infix binary one...

File: mathics/builtin/assignment.py (hunk 1 above)
Comment (mmatera) [inline mathics/builtin/assignment.py]: Now it calls Doc().__str__() from mathics.doc.doc to parse the docstrings.

File: mathics/main.py (hunk 6 above)
Comment (mmatera) [inline mathics/main.py]: I moved this chunk of code to a new branch.

File: mathics/builtin/assignment.py (hunk 2 above)
Comment (sn6uv) [inline mathics/builtin/assignment.py]: Not quite, it could still have a higher precedence than `*` just the second example has no possibility of a unary `Times`.

File: mathics/core/parser/parser.py [hunk 10]
```diff
@@ -236,6 +236,12 @@ def parse_postfix(self, expr1, token, p):
     # Called with one Token and return a Node.
     # Used for prefix operators and brackets.
 
+    def p_PatternTest(self,token):
```
Comment (mmatera) [inline mathics/core/parser/parser.py]: This point was confusing for me: the name of the token has nothing to do with the prefix op...

File: mathics/core/parser/tokeniser.py (hunk 5 above)
Comment (mmatera) [inline mathics/core/parser/tokeniser.py]: I think I have solved this issue now.

File: mathics/builtin/assignment.py (hunk 8 above)
Comment (sn6uv) [inline mathics/builtin/assignment.py]: I think the point in the docs is (it's unclear) that under the hood `Information` is calling `Definitions` to construct its output.

File: mathics/core/parser/tokeniser.py (hunk 5 above)
Comment (sn6uv) [inline mathics/core/parser/tokeniser.py]: The tokeniser does not distinguish between prefix and binary operators at all, that's the parsers job. For example, `1 + 2` will tokenise as a sequence `[Number, Plus, Number]` but `+3` would just be `[Plus Number]`.

The parser (in `core/parser/parser.py`) knows whether it's at the start of an expression or part-way through.

At the start of an expression it calls `parse_p` for parsing prefix expressions. It looks for a manual rule of the form `p_TOKENNAME` (in this case `p_Plus`). Alternatively it looks for the token in `prefix_ops` (defined in `core/parser/operators.py`) which is handled directly. If neither case applies it emits a syntax message.

In the current master `+3` finds `p_Plus` but `?Sin` can't find either a `p_PatternTest` nor find it in the `prefix_ops` dict so it emits the syntax error message that we currently see.

Does that make sense?

File: mathics/core/parser/parser.py (hunk 10 above)
Comment (sn6uv) [inline mathics/core/parser/parser.py]: Yes, one of the quirks of the parser is that is uses the token name (called `tag` in the code) to construct the Node (see the `parse_p` method). This works well for most cases unless the token is reused across different functions with different names.

File: mathics/core/parser/parser.py [hunk 11]
```diff
@@ -413,6 +419,18 @@ def p_Decrement(self, token):
         q = prefix_ops['PreDecrement']
         return Node('PreDecrement', self.parse_exp(q))
 
+
+    def p_Definition(self, token):
+        self.consume()
+        q = prefix_ops['Definition']
+        return Node('Definition', self.parse_exp(q))
```
Comment (sn6uv) [inline mathics/core/parser/parser.py]: I think we can remove this rule now right? It's been replaced with `p_PatternTest`.

File: mathics/core/parser/parser.py [hunk 12]
```diff
@@ -413,6 +419,18 @@ def p_Decrement(self, token):
...
+
+    def p_Information(self, token):
+        self.consume()
+        q = prefix_ops['Information']
+        return Node('Information', self.parse_exp(q))
```
Comment (sn6uv) [inline mathics/core/parser/parser.py]: what about the `LongForm` option?

**Comment by sn6uv** [comment]
One thing that's missing is the `LongForm` options. I'm not sure what the best way to do this is, probably just construct a sub node e.g. `Node('Rule', Symbol('LongForm'), Symbol('True'))` in the `p_` methods.

File: mathics/builtin/assignment.py (hunk 8 above)
Comment (mmatera) [inline mathics/builtin/assignment.py]: Now ? F calls Information[F, LongForm->False]

**Comment by mmatera** [comment]
Now the operators works as they should, without introduce further conflicts, and Information now accepts the option LongForm.  

The only problem now seems to be in the test strings. For some reason, the tester neglects some spaces in the test string, and then it see a mismatch between the expected and the obtained output. Maybe you can give me a clue about how to fix it.

**Comment by sn6uv** [comment]
It looks like a bunch of the jupyter kernel stuff got committed here.

File: mathics/core/parser/operators.py (hunk 3 above)
Comment (mmatera) [inline mathics/core/parser/operators.py]: Still not sure about this. What's the best? To assign the default precedence or to leave it as it is now? In principle, the last option seems to work.

File: mathics/core/parser/parser.py [hunk 13]
```diff
@@ -459,6 +459,20 @@ def p_Decrement(self, token):
         q = prefix_ops['PreDecrement']
         return Node('PreDecrement', self.parse_exp(q))
 
+    def p_PatternTest(self,token):
+        self.consume()
+        q = prefix_ops['Definition']
+        child = self.parse_exp(q)
+        return Node('Information', child, Node('Rule',Symbol("LongForm"), Symbol("False")))
```
Comment (wolfv) [inline mathics/core/parser/parser.py]: nitpick: some formatting (space after comma)

File: mathics/core/parser/parser.py [hunk 14]
```diff
@@ -459,6 +459,20 @@ def p_Decrement(self, token):
...
+
+    def p_Information(self, token):
+        self.consume()
+        q = prefix_ops['Information']
+        child = self.parse_exp(q)
+        if not child.__class__ is Symbol:
+            raise InvalidSyntaxError()
+        return Node('Information', child, Node('Rule',Symbol("LongForm"), Symbol("True")))
```
Comment (wolfv) [inline mathics/core/parser/parser.py]: same here :)

File: mathics/core/parser/tokeniser.py [hunk 15]
```diff
@@ -228,7 +232,7 @@
     '<': ['Get', 'StringJoin', 'LessEqual', 'Less'],
     '=': ['SameQ', 'UnsameQ', 'Equal', 'Unset', 'Set'],
     '>': ['PutAppend', 'Put', 'GreaterEqual', 'Greater'],
-    '?': ['PatternTest'],
+    '?': ['PatternTest','Definition','Information'],
```
Comment (wolfv) [inline mathics/core/parser/tokeniser.py]: spaces

**Comment by wolfv** [comment]
let me know once this is done!

**Comment by mmatera** [comment]
Now I am trying to decouple the patch from other changes that had the
version of Mathics over which I had implemented it. I tell you as soon I
have finished.

File: mathics/core/parser/parser.py (hunk 13 above)
Comment (mmatera) [reply mathics/core/parser/parser.py]: I think all my changes now are mostly pep8 compliant, excepting for a few long lines.

File: mathics/builtin/assignment.py (hunk 2 above)
Comment (mmatera) [reply mathics/builtin/assignment.py]: Now, it parses ? and ?? before + and *:

In[1]:= Hold[??a + b]
Out[1]= Hold[Null + b]

In[2]:= Hold[?a + b]
Out[2]= Hold[Null + b]

In[3]:= Hold[?? a * b]
Out[3]= Hold[Null b]

In[4]:= ? a
Out[4]= Null

I do not know if this current behavior makes more or less sense than the wolfram's

In[1]:= Hold[??a + b]'
Out[1]= Hold[Information[a, LongForm -> True]*Plus[b]]

In[2]:= Hold[?a + b]'
Out[2]= Hold[Information[a, LongForm -> False]*Plus[b]]

 In[3]:= Hold[?? a * b]
 Out[3]= Hold[Information[a * b, LongForm -> True]]

In[4]:= ? a

Information::notfound: Symbol a not found.


Notice that Wolfram's does not show Null Symbol as a return, so actually Information returns Null:

$ wolframscript -c "? Plus"
x + y + z represents a sum of terms. 
Null

Anyway, any of these behaviors should be fine, if there is no practical context in which they will be used...

File: mathics/core/parser/tokeniser.py (hunk 4 above)
Comment (mmatera) [reply mathics/core/parser/tokeniser.py]: In[1]:= a ? b//FullForm
Out[1]= PatternTest[a, b]

In[2]:= ? a b //FullForm
Out[2]= Times[b, Information[a, Rule[LongForm, False]]]


In Wolfram's

 In[1]:= a ? b //FullForm
 Out[1]//FullForm= FullForm[a?b]

 In[2]:= ? a b //FullForm
Global`a
Global`b 
 Out[2]//FullForm= FullForm[Null]

File: mathics/core/parser/tokeniser.py (hunk 5 above)
Comment (mmatera) [reply mathics/core/parser/tokeniser.py]: Done

File: mathics/core/parser/operators.py (hunk 3 above)
Comment (mmatera) [reply mathics/core/parser/operators.py]: Seems to work as it is... we can try to change it to behave more like the weird way Wolfram's does it in weird cases, but at a practical level, it seems that does not matter.

File: mathics/builtin/assignment.py (hunk 7 above)
Comment (mmatera) [reply mathics/builtin/assignment.py]: Done

File: mathics/builtin/inout.py (hunk 9 above)
Comment (mmatera) [reply mathics/builtin/inout.py]: Done

File: mathics/main.py (hunk 6 above)
Comment (mmatera) [reply mathics/main.py]: Done

File: mathics/core/parser/tokeniser.py (hunk 5 above)
Comment (mmatera) [reply mathics/core/parser/tokeniser.py]: Fixed

File: mathics/builtin/assignment.py (hunk 2 above)
Comment (mmatera) [reply mathics/builtin/assignment.py]: Indeed,

Mathics:

In[1]:= Hold[?? a + b] //FullForm
Out[1]= Hold[Plus[Information[a, Rule[LongForm, True]], b]]

In[2]:= Hold[? a + b]//FullForm
Out[2]= Hold[Plus[Information[a, Rule[LongForm, False]], b]]

In[3]:= Hold[?? a * b] //FullForm
Out[3]= Hold[Times[Information[a, Rule[LongForm, True]], b]]

In Wolfram's 
In[1]:= Hold[?? a + b] //FullForm
Out[1]//FullForm= FullForm[Hold[Information[a, LongForm -> True]*Plus[b]]]

In[2]:= Hold[? a + b]//FullForm
Out[2]//FullForm= FullForm[Hold[Information[a, LongForm -> False]*Plus[b]]]

In[3]:=  Hold[?? a * b] //FullForm
Out[3]//FullForm= FullForm[Hold[Information[a * b, LongForm -> True]]]


I don't know if it is really relevant to fix the precedence to reproduce this weird behavior.

File: mathics/core/parser/operators.py (hunk 3 above)
Comment (mmatera) [reply mathics/core/parser/operators.py]: It works as it is

**Comment by mmatera** [comment]
I think now it is ready for merging.

**Comment by mmatera** [comment]
Done. And also I fixed the precedence of ? and ?? operators according to
wolfram's specification (=670).

M

**Comment by GarkGarcia** [comment]
Great, it passed all tests! @mmatera Have the issues pointed out by @wolfv been addressed?

**Comment by mmatera** [comment]
I guess I did...
